src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+       * If the object crosses behind the viewer but not in front, transform 
+         the spherical angles so they are coming from the back of the object,
+         while noting that this range is centered on the back of the object.
+       * If it crosses both, we do not optimize the amount of rays sent.
+
+    5. Compute the intersection of the optimizated range from step 4 and
+       the viewAngles range, accounting for where the optimization range is centered. 
+       If it is empty, the object cannot be visible. If it is not empty, shoot rays at 
+       the desired density in the intersection region. Keep all rays that intersect
+       the object (candidate rays).
+    6. If there are no candidate rays, the object is not visible.
+    7. For each occluding object in occludingObjects: check if any candidate rays 
+       intersect the occluding object at a distance less than the distance they intersected
+       the target object. If they do, remove them from the candidate rays.
+    8. If any candidate rays remain, the object is visible. If not, it is occluded
+       and not visible.
+
+    For visibility of Points/OrientedPoints:
+
+    1. Check if distance from the viewer to the point is greater than visibleDistance.
+       If so, the point cannot be visible
+    2. Create a single candidate ray, using the vector from the viewer to the target.
+       If this ray is outside of the bounds of viewAngles, the point cannot be visible.
+    3. For each occluding object in occludingObjets: check if the candidate ray hits
+       the occluding object at a distance less than the distance from the viewer to the
+       target point. If so, then the object is not visible. Otherwise, the object is visible.
+
+    Args:
+        position: Position of the viewer, accounting for any offsets.
+        orientation: Orientation of the viewer.
+        visibleDistance: The maximum distance the viewer can view objects from.
+        viewAngles: The horizontal and vertical view angles, in radians, of the viewer.
+        rayDensity: The density of rays used in visibility calculations. The total number
+          of rays sent will be (rayDensity**2) times the area the total view angle area being
+          evaluated (after optimization).
+        target: The target being viewed. Currently supports Point, OrientedPoint, and Object.
+        occludingObjects: An optional list of objects which can occlude the target.
+    """
+    occludingObjects = list(occludingObjects)
+
+    if isinstance(target, (Region, Object)):
+        # Extract the target region from the object or region.
+        if isinstance(target, Region):
+            raise NotImplementedError()
+        elif isinstance(target, (Object)):
+            target_region = target.occupiedSpace
+
+            # If the object contains its center and we can see the center, the object
+            # is visible.
+            if target_region.containsPoint(target.position) and \
```

Comment:
This condition seems arbitrary. I'd suggest having all mesh regions provide a `representativePoint` which can be the center if it's contained and otherwise an arbitrary contained point.